Document value($X) in message (#716)
returntocorp/semgrep-docs: docs/writing-rules/pattern-syntax.mdx

foo(1,2,3,1,2)
```

### Metavariables in messages

You can use metavariables in messages. To do this, simply add the metavariable in the message (e.g. `Found $X!`) and it will be replaced with the value matched by the metavariable.

For example, let's say your message is 

`Creating a buffer using $X`

and your formula is

```yaml
- pattern: |
   byte[] buf = new byte[$X];
```

You run this against

```java
byte[] buf = new byte[size];
```

The message you get is `Creating a buffer using size`.

Sometimes, however, the value matched by the metavariable is not the real value you were looking for. Consider this slightly different formula and program:

Formula:

```yaml
- patterns:
   - pattern: byte[] buf = new byte[$X];
   - metavariable-comparison:
        metavariable: $X
        comparison: $X < 2048
```

Program:

```java
int size = 512;
byte[] buf = new byte[size];
```

These will match because Semgrep performs constant propagation. It knows that `size` is `512`, so it knows that a buffer is created with size less than `2048`. But what is `$X`?

> If we use the message `Creating a buffer using $X` again, we see that the size is `size`. This is because the value matched by `$X` within the code is `size`. However, we know that the underlying value of `size` is `512`, and we might want to access that in our message. To do that, we can use `value($X)`. `value($X)` will be replaced by the propagated value of a metavariable if Semgrep has propagated a value, or just the metavariable otherwise. 

Author: I think B is right :D So if there's a propagated value to use, then `value($X)` will be replaced with the propagated value. Otherwise, `value($X)` will be replaced with just the metavariable match. But I replied to another comment explaining that this isn't always desirable

> These will match because Semgrep performs constant propagation. It knows that `size` is `512`, so it knows that a buffer is created with size less than `2048`. But what is `$X`?
> 
> If we use the message `Creating a buffer using $X` again, we see that the size is `size`. This is because the value matched by `$X` within the code is `size`. However, we know that the underlying value of `size` is `512`, and we might want to access that in our message. To do that, we can use `value($X)`. `value($X)` will be replaced by the propagated value of a metavariable if Semgrep has propagated a value, or just the metavariable otherwise. 

Reviewer: @emjin understand! So is this correct? Feel free to make little edits if you find them necessary.

```suggestion
Semgrep matches this code using constant propagation. Semgrep recognizes that `size` is `512`. Consequently, Semgrep evaluates that the size of the buffer is less than `2048`. But what is the value of `$X`?

If the rule message states `Creating a buffer using $X`, the resulting message output is not helpful in this particular case: `Creating a buffer using size`. This is caused by the value of `$X` within the code, which is `size`. However, the underlying value of `size` is `512`. The goal of the rule message is to access this underlying value in our message.

To retrieve the correct value in the case described above, use `value($X)` in the rule message (
(for example `Creating a buffer using value($X)`). Semgrep replaces the `value($X)` with the underlying propagated value of a metavariable.
```

> If we use the message `Creating a buffer using $X` again, we see that the size is `size`. This is because the value matched by `$X` within the code is `size`. However, we know that the underlying value of `size` is `512`, and we might want to access that in our message. To do that, we can use `value($X)`. `value($X)` will be replaced by the propagated value of a metavariable if Semgrep has propagated a value, or just the metavariable otherwise. 

Reviewer: Thank you for the explanation here! :-)

```

The resulting message is:

```
Creating a buffer using size
```

Semgrep substitutes the $X with the value detected within the metavariable. In this case, the value was `size`.

However, in some cases, the matched value of the metavariable is not the real value you were looking for. See the following formula and testing code:

Part of a Semgrep rule (formula):

```yaml
- patterns:
   - pattern: byte[] buf = new byte[$X];
   - metavariable-comparison:
        metavariable: $X
        comparison: $X < 2048
```

Testing code:

```java
int size = 512;
byte[] buf = new byte[size];
```

Semgrep matches this code because it performs constant propagation. Therefore, Semgrep recognizes that the value of `size` is `512`. Consequently, Semgrep evaluates that the buffer size is less than `2048`. But what is the value of `$X`?

If the rule message states `Creating a buffer using $X`, the resulting message output is not helpful in this particular case: `Creating a buffer using size`. This is caused by the value of `$X` within the code, which is `size`. However, the underlying value of `size` is `512`. The goal of the rule message is to access this underlying value in our message.

> To retrieve the correct value in the case described above, use `value($X)` in the rule message (for example (`Creating a buffer using value($X)`). Semgrep replaces the `value($X)` with the underlying propagated value of a metavariable.

Author: ```suggestion
To retrieve the correct value in the case described above, use `value($X)` in the rule message (for example (`Creating a buffer using value($X)`). Semgrep replaces the `value($X)` with the underlying propagated value of the metavariable `$X` if it computes one (otherwise, it will just use the matched value).
```

Run the following example in Semgrep Playground to see the message (click **Open in Editor**, and then **Run**, unroll the **1 Match** to see the message):

<iframe title="Metavariable value in message example" src="https://semgrep.dev/embed/editor?snippet=returntocorp:value-in-message-example" width="100%" height="432" frameborder="0"></iframe>

> 

Reviewer: ```suggestion
```
